crates/resolver/src/name_server/name_server.rs

```diff
@@ -219,6 +219,62 @@ impl<P: ConnectionProvider> NameServerState<P> {
 
 struct ConnectionMeta {
     status: AtomicU8,
+    srtt: DecayingSrtt,
+}
+
+impl ConnectionMeta {
+    fn new(initial_srtt: Duration) -> Self {
+        Self {
+            status: AtomicU8::new(Status::Init.into()),
+            srtt: DecayingSrtt::new(initial_srtt),
+        }
+    }
+
+    /// Records the measured `rtt` for a particular result.
+    ///
+    /// Tries to guess if the result was a failure that should penalize the expected RTT.
+    fn record(&self, rtt: Duration, result: &Result<DnsResponse, ProtoError>) {
+        let error = match result {
+            Ok(_) => {
+                self.srtt.record(rtt);
+                return;
+            }
+            Err(err) => err,
+        };
+
+        use ResponseCode::*;
+        match error.kind() {
+            ProtoErrorKind::NoRecordsFound(NoRecords { response_code, .. }) => {
+                match response_code {
+                    ServFail => self.srtt.record_failure(),
+                    _ => self.srtt.record(rtt),
+                }
+            }
+            ProtoErrorKind::Busy | ProtoErrorKind::Io(_) | ProtoErrorKind::Timeout => {
+                self.srtt.record_failure()
+            }
+            #[cfg(feature = "__quic")]
+            ProtoErrorKind::QuinnConfigError(_)
+            | ProtoErrorKind::QuinnConnect(_)
+            | ProtoErrorKind::QuinnConnection(_)
+            | ProtoErrorKind::QuinnTlsConfigError(_) => self.srtt.record_failure(),
+            #[cfg(feature = "__tls")]
+            ProtoErrorKind::RustlsError(_) => self.srtt.record_failure(),
+            _ => {}
+        }
+    }
+}
+
+impl Default for ConnectionMeta {
+    fn default() -> Self {
+        // Initialize the SRTT to a randomly generated value that represents a
+        // very low RTT. Such a value helps ensure that each server is attempted
+        // early.
+        Self::new(Duration::from_micros(rand::random_range(1..32)))
```

Comment:
Maybe `DecayingSrtt` should have its own `impl Default` that calls its constructor with the rand `Duration`? 

It feels like a bit too much indirection to have it happening in the `ConnectionMeta` default, and nothing else calls `ConnectionMeta::new()` with a custom `Duration` AFAICT.